orttraining/orttraining/models/runner/training_runner.cc

```diff
@@ -213,6 +212,11 @@ Status TrainingRunner::Initialize() {
   // Retrieve pipeline information from configuration result.
   VectorString fetch_names;
   if (params_.pipeline_parallel_size > 1) {
+    // Instead of inside constructor, we initialize pipeline_schedule_ and pipeline_worker_pool_ here
+    // because they dependent on the result of TraningSession::ConfigureForTraining(...).
```

Comment:
>TraningSession::ConfigureForTraining [](start = 47, length = 36)

Not sure why they would have a dependency on TraningSession::ConfigureForTraining. The initialization should be purely depending on training parameter `gradient_accumulation_steps` and `pipeline_parallel_size`?

Reply:
The following piece of code needs to be executed before creating pipeline schedule.
```
  DistributedRunContext::CreateInstance({config.distributed_config.world_rank,
                                         config.distributed_config.world_size,
                                         config.distributed_config.local_rank,
                                         config.distributed_config.local_size,
                                         config.distributed_config.data_parallel_size,
                                         config.distributed_config.horizontal_parallel_size,
                                         config.distributed_config.pipeline_parallel_size});

```

---
In reply to: [538715364](https://github.com/microsoft/onnxruntime/pull/5815#discussion_r538715364) [](ancestors = 538715364)